text/0000-natvis.md

```diff
@@ -0,0 +1,428 @@
+`do_some_work()`. (Perhaps some code path has triggered a panic.) Carol wants
+to look at the state of the app, inside the `find_c_defines` function,
+and she specifically wants to see what the state of `captures` is. So she
+selects the `find_c_defines` call frame and looks at the local variables
+window.
+
+Unfortunately, the debugger's view of the `captures` variable does not give
+her any useful information at all. It shows only something like:
+
+```text
+> Variables
+  > captures: {...}
+    > text: "...the entire input text..."
+    > locs: {...}
+      > __0: (4) vec![None, None, None, None]
+      > named_groups: (refs:2) size=0, capacity=1
+        > [raw]: alloc::sync::Arc<std::collections::hash::map::HashMap<...>>
+          > ptr: {pointer:0xNNNNNNNN}
+            > pointer: {...}
+              > strong: {...}
+              > weak: {...}
+              > data: size=0, capacity=1
+                > base: {...}
+                  > hash_builder: {...}
+        ...
+```
+
+The debugger shows the structure of the data, not its meaning. It is useless.
+Even the implementor of `regex` would have a hard time knowing how to decode
+this. In reality, when trying to understand the state of the captures
+variable there are several methods defined for a `Captures` type that paint
+the actual picture in terms of the information a Rust developer would like to
+extract from this variable. In order to meaningfully understand what the `Captures`
+type is truly trying to tell us, it would be very helpful to visualize this data
+differently in the debugger.
+
+What we _want_ is something like this:
+
+```text
+> Variables:
+  > captures: {...}
+    > $1: "SOME_CONSTANT"
+    > $2: "42"
+    > $3: "// some developer comment"
+```
+
+This RFC will describe how to support adding Natvis visualizations which is supported by:
+
+* The Windows Debugger (WinDbg)
+* Visual Studio Debugger.
+
+It should be easy for crate developers to add debugger visualizations to their
+crates.
+
+## Supporting Natvis
+
+This section describes how Microsoft's Natvis is supported in Rust.
+
+To use Natvis, developers write XML documents that describe how debugger types
+should be displayed. (See: https://docs.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects?view=vs-2019)
+The Natvis XML files provide patterns, which match type names, and for matching
+types, a description of how to display those types. This allows for some
+limited support for generic types.
+
+When writing Natvis files for C++, developers write a standalone XML document
+and add it to their project. The build system (such as Visual Studio) knows
+how to package the Natvis file into the debug data (the PDB file) for the
+project, and the debugger knows how to find the Natvis XML in each PDB file.
+
+Developers can add one or more standalone Natvis XML files to their crate.
+The `Cargo.toml` file specifies the name of these Natvis files. This is the
```

Comment:
Is only the name specified? What does the declaration look like? If we only need the name, why not list the contents of a directory?

Reply:
The path to the .natvis file would be specified. I'll update the language to reflect this. Currently this RFC is aimed at listing files. Presumably in the future this could be updated to include support for directories but just as the auto-discovering of natvis files it's out of scope for this initial RFC.